### PR TITLE
Release automation improvements

### PR DESCRIPTION
**Pull request by justinas**
This PR tries to address https://github.com/gravitational/cloud/issues/1199 and https://github.com/gravitational/cloud/issues/1272 .

There are several changes included under this single PR, in order to cut down on the amount of backports and merge conflicts. Each commit includes an extended message where I tried to include some background information. Please do not hesitate to raise any questions or concerns. 

This should be the penultimate batch of changes concerning integration with the release service. After this is merged, if no new issues occur, we shall follow up with another PR that switches to the production endpoint and removes `failure:ignore` from these steps.

Tag pipeline test: https://drone.platform.teleport.sh/gravitational/teleport/11278

### CODE REVIEW

**Comment by russjones** [comment]
@justinas Just a heads up, we might be slow on reviews this week due to focusing on getting Teleport 9 out, but we'll try and get to this.

@r0mant @fheinecke @tcsc Can you review?

File: dronegen/tag.go
Comment (tcsc) [reply dronegen/tag.go:350]: Who receives that `POST`, btw?

File: dronegen/tag.go
Comment (tcsc) [inline dronegen/tag.go:354]: Why is status 409 special here?

**Review by tcsc — APPROVED** [review]
Looks OK to me, modulo a few questions. You will want to get someone with better `bash` than me to review it before submission, though ;-)

File: dronegen/common.go
Comment (justinas) [reply dronegen/common.go:139]: Mostly this was an attempt to keep the descriptions as close to the current ones. I do agree that it makes sense to be more explicit in some cases.

For darwin, bitness is not relevant, but we could add an arch specifier, e.g.:
* MacOS Intel
* MacOS Apple Silicon

File: dronegen/tag.go
Comment (justinas) [reply dronegen/tag.go:481]: I'm not sure I understand the comment. The indentation seems to be unchanged on this line?

File: dronegen/tag.go
Comment (justinas) [reply dronegen/tag.go:350]: This is all handled by "releaseserver" https://github.com/gravitational/cloud/tree/master/pkg/releaseserver . This request in particular handled here https://github.com/gravitational/cloud/blob/1e466d0df1a3aef70690e99a0e3cabb0e8c5b602/pkg/releaseserver/handler/handler.go#L36-L37

File: dronegen/common.go
Comment (r0mant) [inline dronegen/common.go:111]: I don't know if we want to panic. Can we return a proper error and include the actual `b.os` value in the error message for troubleshooting?

File: dronegen/common.go
Comment (r0mant) [inline dronegen/common.go:130]: Same comment as above.

File: dronegen/tag.go
Comment (r0mant) [reply dronegen/tag.go:49]: Instead of using even more bash, can we please extend our [version check tool](https://github.com/gravitational/teleport/tree/v9.0.1/build.assets/tooling/cmd/check) to support this and use it instead? It's already used in some other places e.g. [checking for pre-release](https://github.com/gravitational/teleport/blob/v9.0.1/.drone.yml#L4894).

File: dronegen/tag.go
Comment (r0mant) [inline dronegen/tag.go:342]: Any chance all of this can be implemented in Go? I'm having a hard time following the logic here. I think we should be reducing amount of bash code we have, not adding it.

**Review by r0mant — CHANGES_REQUESTED** [review]

File: dronegen/common.go
Comment (justinas) [reply dronegen/common.go:111]: We could return an error, but that introduces a lot of error propagation up the call stack (unless at some point we still `panic()` or `log.Fatal()`):
* `(*buildType).Description()`
* `tagCreateReleaseAssetCommands()`
* `tagPipeline()`
* `tagPipelines()`
* `main()`

All for what is currently the sole fallible step in generating the YAML itself.

I agree with improving the error message.

File: dronegen/tag.go
Comment (justinas) [reply dronegen/tag.go:342]: This is a good idea, we have recently created a basic Go client (see https://github.com/gravitational/cloud/pull/1433), but currently it implements a different subset of operations on this API. I would definitely like to move as much logic as possible to the Go client in the long run (it would also have the benefit of reducing backports to Teleport version branches), but I'm not sure if it is doable in Q1 (by the end of which we would like for CDN to go live).

File: dronegen/tag.go
Comment (justinas) [reply dronegen/tag.go:49]: `cmd/check` does not seem to exist in branch/v7, and does not seem to be fully backported (up-to-date) in branch/v8. We could try to backport it everywhere, perhaps.

File: dronegen/tag.go
Comment (justinas) [reply dronegen/tag.go:342]: Created an issue so this does not get lost: https://github.com/gravitational/cloud/issues/1504

**Comment by justinas** [comment]
Some merge conflicts arose in `.drone.yml` (as they usually do). Will solve those and re-run dronegen after getting the approval.

**Review by r0mant — APPROVED** [review]
@justinas I'm gonna approve cause I don't want to block you but can you please make sure that https://github.com/gravitational/cloud/issues/1504 does not get lost and is actually addressed? For all future changes to the pipeline, unless it's something trivial, please prefer writing tools in Go as well. cc @wadells

File: dronegen/tag.go
Comment (wadells) [reply dronegen/tag.go:49]: > `cmd/check` does not seem to exist in branch/v7, and does not seem to be fully backported (up-to-date) in branch/v8. We could try to backport it everywhere, perhaps.

I suspect https://github.com/gravitational/teleport/pull/10436 is where the port was not fully backported.

> I chose not to backport the docker image publishing fixes (for https://github.com/gravitational/teleport/issues/9494), as that drone pipeline is only ever run off master.

If it helps these efforts, I can backport the rest of https://github.com/gravitational/teleport/pull/10295 to v8 and v7.  Let me know.

**Pull request merged**
